geps/gep-1619.md

```diff
@@ -163,43 +268,384 @@ Java application servers such as Tomcat and Jetty, were the first to standardize
+configurations, like allowing customization of the cookie name, could prevent certain implementations from conforming to
+the spec. In other words, opting for a higher-level API provides better interoperability among our implementations.
+
+#### Target Persona
+
+Referring to the [Gateway API Security Model](https://gateway-api.sigs.k8s.io/concepts/security-model/#roles-and-personas),
+the target kubernetes role/persona for session persistence are application developers, as mentioned in the [When does an application require session persistence?](#when-does-an-application-require-session-persistence)
+section. It is the responsibility of the application developers to adjust the persistence configuration to ensure the
+functionality of their applications.
+
+#### Prior Art
+
+Referring to our [Implementations](#Implementations) tables on session persistence, the majority of Gateway API
+implementations designed session persistence in their APIs to be attached to a service or backends. This should be
+considered cautiously, as making associations to Gateway API's notion of Gateway, Route, and Service to other
+implementation's objects is hard to directly translate. The idea of a route in Gateway API is often not the same as a
+route in any given implementation.
+
+#### Metaresource Policy Design
+
+In order to apply session persistence configuration to both a service and a route, we will implement it as a
+metaresource policy, as outlined in [Policy Attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment/#direct-policy-attachment).
+The metaresource is named `SessionPersistencePolicy` and is only responsible for configuring session persistence for services.
+It is defined as a [Direct Policy Attachment](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/#direct-policy-attachment)
+without defaults or overrides, applied to the service it is targeting.
+
+Attaching the `SessionPersistencePolicy` metaresource to a service will be a core support level feature, while attaching
+it to a route will be considered extended. This distinction arises because most existing implementations primarily
+support attaching to a service, while attaching to a route is less common and involves greater complexities (see [API Attachment Point: Route vs. Service](#api-attachment-point-route-vs-service)
+for more details).
+
+#### API Attachment Point: Route vs. Service
+
+Session persistence augments both the concepts of routes and services. However, the functionality may vary slightly
+depending on the object to which it is attached. It is important for implementations to understand these differences 
+when implementing the session persistence API. 
+
+##### Service
+
+A Kubernetes service is an abstract representation of a set of pods that work together to provide a specific
+functionality. Services perform load balancing, routing traffic from a service endpoint to a pod. Therefore, in its
```

Comment:
It is strange to not mention that K8S Service also provides a persistent session behavior at L4, via sessionAffinity=ClientIP.  If Cookie-based persistence is defined, it will take priority over the L4 sessionAffinity.

Reply:
Well, I'm actually wrong - Cookie persistence will be used when requests to the backend are done via the Gateway and HttpRoute. 

sessionAffinity=ClientIP should still be used ( and we should strongly recommend that ) for all in-cluster traffic that uses the K8S Service directly. There is no guarantee all http requests are through the gateway IP, and if the app is designed to expect persistence - both paths must be persistent.

Reply:
> Well, I'm actually wrong - Cookie persistence will be used when requests to the backend are done via the Gateway and HttpRoute.

Could you expand on why persistence will be used when requests to the backend are going through GWAPI objects? Sorry, fundamental GWAPI question, but don't most, if not all, implementations use the service just to look up a group of backends? And then encode the backends IPs and ports into their proxy config? I.e. do implementation ever use the Cluster IP of the service? Is that why?

I've got a todo on K8S service relationship with this GEP, but I can go ahead and add a blurb about it.

Reply:
You are (unfortunately) right - most gateway implementations ignore the Service sessionAffinity, and for good reasons: it wouldn't make sense for the gateway client IP to always go to one backend. That's one of the reasons a gateway that naively sends to the ClusterIP of the service is likely to be broken when a Service has that setting. 

My point was that if a backend is designed for affinity - and it has internal K8S clients that call it directly, using the  sessionsvc.ns.svc.cluster.local name - they need to configure the Service sessionAffinity for those internal clients. 

As a side note - while Istio gateways (and others) generally treat Service as bag of endpoints and LB between them, ignoring the ClusterIP - there are APIs and corner cases where we don't do this but fallback to 'original dst', and if the service has K8S Service.sessionAffinity will still be ok for sidecars - but not ok for Gateways or Waypoints. Current Istio does break Service.sessionAffinity - but has a way ( which I forgot...) to force use of ClusterIP where it would work again. Locality (sending only to pods on same node) had similar problems but now is supported.